Fix repeat non owning

maciejkula opened the pull request:
Tentative solution for https://github.com/pytorch/pytorch/issues/4054.

I must admit that the actual details of what is happening in `tensor.repeat` remain somewhat opaque to me, so I opted for the minimal possible change.

I would be happy to add some more robust tests; have you looked into [Hypothesis](https://hypothesis.readthedocs.io/en/latest/)? It's probably the most advanced property-based testing library out there, and is _extremely_ good at making it easier to write effective tests.

maciejkula committed:
Add addiitonal tensor.repeat tests.

These test tensor.repeat when using numpy-shared storage. The new
test adds some additional test cases and compares against np.tile.

maciejkula committed:
Make sure tensor.repeat works for tensors with shared storage.

Clone the input tensor instead of resizing it in place.

albanD commented on torch/tensor.py:
```diff
@@ -274,7 +274,7 @@ def repeat(self, *sizes):
             raise ValueError('Number of dimensions of repeat dims can not be '
                              'smaller than number of dimensions of tensor')
 
-        xtensor = src.new().set_(src)
```
Hi,

I don't think you want to clone here, this is not needed,
A minimal fix would be to change the `resize_` call l284 by a `view` (and remove line 290 that also contains a `resize_` and that is redundant with l284).

maciejkula replied:
I may have to read the cpp sources to figure out what these do.

maciejkula committed:
Change clone to view.

maciejkula commented:
Updated to use `.view`.

apaszke commented on torch/tensor.py:
```diff
@@ -280,14 +280,13 @@ def repeat(self, *sizes):
 
         size = torch.Size([a * b for a, b in zip(xsize, repeats)])
-        xtensor.resize_(torch.Size(xsize))
+        xtensor = xtensor.view(torch.Size(xsize))
```
I think this will break if the input tensor is non-contiguous.

apaszke replied:
If the only change this line does is unsqueezeing some dimensions then let’s use unsqueeze instead of view (it works will all kinds of tensors)

maciejkula committed:
Clean up tensor.repeat. Add noncontiguous input tests.

I added some explanatory comments to the function itself and added a
test that repeats a non-contiguous input.

maciejkula commented:
@apaszke I moved to using an `expand` to add the new leading dimensions, and added a test for repeating non-contiguous inputs.

apaszke commented on torch/tensor.py:
```diff
@@ -266,30 +266,31 @@ def repeat(self, *sizes):
-        xxtensor = xtensor.expand_as(urtensor)
-        urtensor.copy_(xxtensor)
+
+        urtensor.copy_(xtensor.expand_as(urtensor))
```
Is this `expand_as` really needed? I'm not 100% how these unfolds let us achieve repeating, but it seems that there's no guarantee that `xtensor` is contiguous, so if this wasn't a no-op I would expect it to break.

maciejkula replied:
To be perfectly frank I have no idea what is going on there. I think `urtensor` gets some new dimensions beyond `xtensor`, then the copy semantics somehow make it work after expanding.

apaszke commented on test/test_torch.py:
```diff
@@ -3873,18 +3873,49 @@ def test_expand(self):
-        result = tensor.repeat(torchSize)
-        self.assertEqual(result.size(), target, 'Error in repeat using result and LongStorage')
-        self.assertEqual(result.mean(0).view(8, 4), tensor, 'Error in repeat (not equal)')
+
```
Actually can you please make the new parts of this test a separate function, and use `skipIf` to disable it if numpy is not available?

apaszke reviewed (commented):
Looks good. Some minor comments, and should be good to merge once they are resolved.

maciejkula committed:
Skip numpy-based repeat tests if numpy not found.

apaszke commented:
Thank you!
